service/storage_proxy.cc

```diff
@@ -433,51 +535,72 @@ class storage_proxy::remote {
         errors_info errors;
         ++p->get_stats().received_mutations;
         p->get_stats().forwarded_mutations += forward.size();
-        co_await coroutine::all(
-            [&] () -> future<> {
-                try {
-                    // FIXME: get_schema_for_write() doesn't timeout
-                    schema_ptr s = co_await get_schema_for_write(schema_version, netw::messaging_service::msg_addr{reply_to, shard});
-                    // Note: blocks due to execution_stage in replica::database::apply()
-                    co_await apply_fn(p, trace_state_ptr, std::move(s), m, timeout);
-                    // We wait for send_mutation_done to complete, otherwise, if reply_to is busy, we will accumulate
-                    // lots of unsent responses, which can OOM our shard.
-                    //
-                    // Usually we will return immediately, since this work only involves appending data to the connection
-                    // send buffer.
-                    auto f = co_await coroutine::as_future(send_mutation_done(netw::messaging_service::msg_addr{reply_to, shard}, trace_state_ptr,
+
+        const auto reply_to_addr = netw::messaging_service::msg_addr{reply_to, shard};
+
+        if (fence) {
+            const auto partition_key = [&]() {
+                if constexpr(std::is_same_v<decltype(in), frozen_mutation>) {
+                    return in.key();
+                } else {
+                    return in.update.key();
+                }
+            }();
+            auto stale_topology = co_await try_apply_fence(*fence, timeout,
+                write_request_info{*this, schema_version, reply_to_addr, partition_key});
+            if (stale_topology) {
+                errors.count += (forward.size() + 1);
+                errors.local = std::move(*stale_topology);
+            }
+        }
+
+        if (errors.count == 0) {
+            co_await coroutine::all(
+                [&] () -> future<> {
+                    try {
+                        // FIXME: get_schema_for_write() doesn't timeout
+                        auto s = co_await get_schema_for_write(schema_version, reply_to_addr);
```

Comment:
Please do not replace concrete type with an auto. If this would have been a new code I wouldn't complain, but why erase the information we have with no reason?

Reply:
ok